File: examples/async/lighter_ws_private_readonly.py
"""Open a Lighter private WebSocket user-data stream."""

import asyncio

from dcex.lighter import Network
from dcex.ws import lighter


async def read_network(network: Network) -> None:
    """Read one private order event from a Lighter network."""
    async with lighter.private_from_env(network=network) as ws:
        await ws.subscribe_account_all_orders()
        print(network.value, await ws.recv())


async def main() -> None:
    """Open Lighter and Robinhood private streams concurrently."""
    await asyncio.gather(
        read_network(Network.MAINNET),
        read_network(Network.ROBINHOOD),
    )


if __name__ == "__main__":
    asyncio.run(main())

File: crates/dcex/examples/lighter_ws_private_readonly.rs
use std::time::Duration;

use dcex::exchanges::lighter::LighterNetwork;
use dcex::ws::lighter::LighterPrivateWebSocket;

#[tokio::main]
async fn main() -> dcex::Result<()> {
    tokio::try_join!(
        read_network(LighterNetwork::Mainnet),
        read_network(LighterNetwork::Robinhood),
    )?;
    Ok(())
}

async fn read_network(network: LighterNetwork) -> dcex::Result<()> {
    let mut ws = LighterPrivateWebSocket::with_env_credentials(network, Duration::from_secs(10))?;
    ws.connect().await?;
    ws.subscribe_account_all_orders().await?;
    println!("{network}: {}", ws.recv().await?);
    ws.close().await?;
    Ok(())
}
